src/CrossChainReceiverFactory.sol

```diff
@@ -0,0 +1,406 @@
+    // @inheritdoc IERC1271
+    function isValidSignature(bytes32 hash, bytes calldata signature)
+        external
+        view
+        override
+        /* `_verifyRoot` hashes `_cachedThis`, making this function implicitly `onlyProxy` */
+        returns (bytes4)
+    {
+        { // Merkle proof validation
+            address owner;
+            bool validOwner;
+            bytes32 leaf;
+            bytes32[] calldata proof;
+
+            assembly ("memory-safe") {
+                // This assembly block is equivalent to:
+                //     leaf = keccak256(abi.encode(hash, block.chainid));
+                // except that it's cheaper and doesn't allocate memory. We make the assumption here that
+                // `block.chainid` cannot alias a valid tree node or signing hash. Realistically,
+                // `block.chainid` cannot exceed 2**53 - 1 or it would cause significant issues elsewhere in
+                // the ecosystem. This also means that the sort order of the hash and the chainid is
+                // backwards from what `MerkleProofLib` produces, again protecting us against extension
+                // attacks.
+                mstore(0x00, hash)
+                mstore(0x20, chainid())
+                leaf := keccak256(0x00, 0x40)
+
+                // Following assembly blocks are equivalent to:
+                //     (owner, proof) = abi.decode(signature, (address, bytes32[]));
+                // except we omit all the range and overflow checking.
+                owner := calldataload(signature.offset)
+                validOwner := iszero(shr(0xa0, owner))
+            }
+
+            if (validOwner) {
+               assembly ("memory-safe") {
+                    // continuation of previous abi.decode
+                    proof.offset := add(signature.offset, calldataload(add(0x20, signature.offset)))
+                    proof.length := calldataload(proof.offset)
+                    proof.offset := add(0x20, proof.offset)
+               }
+               if (_verifyRoot(MerkleProofLib.getRoot(proof, leaf), owner)) {
+                    return IERC1271.isValidSignature.selector;
+                }
+            }  
+        }
+
+        { // ERC7733 validation
+            // For automatic detection that the smart account supports the nested EIP-712 workflow,
+            // See: https://eips.ethereum.org/EIPS/eip-7739.
+            // If `hash` is `0x7739...7739`, returns `bytes4(0x77390001)`.
+            // The returned number MAY be increased in future ERC7739 versions.
+            unchecked {
+                if (signature.length == uint256(0)) {
+                    // Forces the compiler to optimize for smaller bytecode size.
+                    if (uint256(hash) == ~signature.length / 0xffff * 0x7739) return 0x77390001;
+                }
+            }
+            address owner_ = owner();
+            if ((owner_ != address(0)) && _erc1271IsValidSignature(hash, signature, owner_)) {
```

Comment:
Is the check `owner_ != address(0)` required?

Reply:
kind of a way to skip the whole validation flow but we can remove it if you don't see it is useful